packages/@uppy/core/src/Uppy.ts

```diff
@@ -58,23 +58,86 @@ export type UnknownPlugin<
   PluginState extends Record<string, unknown> = Record<string, unknown>,
 > = BasePlugin<any, M, B, PluginState>
 
-// `OmitFirstArg<typeof someArray>` is the type of the returned value of `someArray.slice(1)`.
-type OmitFirstArg<T> = T extends [any, ...infer U] ? U : never
+/**
+ * ids are always `string`s, except the root folder's id can be `null`
+ */
+export type PartialTreeId = string | null
+
+export type PartialTreeStatusFile = 'checked' | 'unchecked'
+export type PartialTreeStatus = PartialTreeStatusFile | 'partial'
+
+export type PartialTreeFile = {
```

Comment:
I think having `PartialTreeBase` would add a level of indirection without making these types more readable/easier to work with.

In general - I avoided having many types here, because I want people who work on these files to largely remember existing types by heart (that's just 4 types - `PartialTree`, `~File`, `~FolderNode`, `~FolderRoot`).

E.g., one really seductive type was `PartialTreeDisplayedNode = PartialTreeFile | PartialTreeFolderNode` (because these are the only nodes that can be seen in the interface), but I decided against it to cut down on n of types the developer would have to memorize.